### PR TITLE
roachprod: fix update & get latest

### PR DESCRIPTION
**Pull request by srosenberg**
Previously, roachprod binary was fetched directly from
the artifacts of a corresponding TeamCity build.
After moving it to Github actions with Engflow, the
artifact is no longer mapped to a unique URL, thus
making it cumbersome to fetch via an API.

This change updates Roachtest Nightly (in TC for
the foreseeable future) to upload the roachprod binary
into GCS, where we already store other artifacts, e.g.,
(roachperf) stats.json. Consequently, we update
`scripts/roachprod-get-latest.sh` and `roachprod update`
to fetch the binary from the new location. Finally, we
add additional sanity checks around "staleness", including
a warning in roachprod (on startup), if the binary is older
than two weeks.

Note that this PR makes only linux/amd64 roachprod binaries
available. For mac and arm64 builds, we'd have to configure
CI to upload the corresponding binaries.

Informs: #120750
Epic: CRDB-8035
Release note: None

### CODE REVIEW

**Comment by srosenberg** [comment]
Emulated "staleness" check which emits this warning in `roachprod`,

```
roachprod
WARN: roachprod binary is >= 2 weeks old (569h7m14.325s); latest sha: "abcdef"
WARN: Consider updating the binary: `roachprod update`
```

File: build/teamcity/cockroach/nightlies/roachtest_compile_bits.sh
Comment (rickystewart) [inline build/teamcity/cockroach/nightlies/roachtest_compile_bits.sh:5]: As we depend on the variable `$root`, this won't work unless the person running the script has already executed `teamcity-support.sh` in this shell. (And then e.g. ran this script with `. roachtest_compile_bits.sh`) Maybe true in all situations where we use this script, but worth pointing out.

File: build/teamcity/cockroach/nightlies/roachtest_nightly_impl.sh
Comment (rickystewart) [inline build/teamcity/cockroach/nightlies/roachtest_nightly_impl.sh:62]: Mistake?

File: build/teamcity/util/roachtest_util.sh
Comment (rickystewart) [inline build/teamcity/util/roachtest_util.sh:78]: **Optional**: Would it be better if we instead just had one file `$os/$arch/roachprod.$branch.latest`? Or, even just one `latest` binary for `master` (per OS/arch combo)?

We only need the latest one, right? Realistically, people are not going to be downloading `roachprod` for older SHA's. This would also obviate the need for garbage collection on the bucket.

File: scripts/roachprod-get-latest.sh
Comment (rickystewart) [inline scripts/roachprod-get-latest.sh:15]: Is there any garbage collection set up, or is the idea that we would be holding on to all of these roachprods indefinitely? Each is ~100MB. Storage costs are going to start adding up quickly.

File: build/teamcity/util/roachtest_util.sh
Comment (rickystewart) [inline build/teamcity/util/roachtest_util.sh:71]: Since we only run the roachtest nightlies on Linux machines, wouldn't this mean we never publish Mac binaries?

**Review by rickystewart — CHANGES_REQUESTED** [review]

**Comment by rickystewart** [comment]
Separately, lots of lint failures. Consider using `lintonbuild` so you don't have to go back-and-forth with CI :)

**Comment by srosenberg** [comment]
> Consider using `lintonbuild` so you don't have to go back-and-forth with CI :)

I did for a while actually, until it got too slow :) Especially, when you're doing multiple builds (on different branches) back-to-back.

File: build/teamcity/util/roachtest_util.sh
Comment (srosenberg) [reply build/teamcity/util/roachtest_util.sh:71]: Indeed. I don't think it's a deal breaker for this PR, but we may need to find/patch a job that already builds for mac, maybe "local roachtest"?

File: scripts/roachprod-get-latest.sh
Comment (srosenberg) [reply scripts/roachprod-get-latest.sh:15]: We could configure a retainment policy on the bucket... but considering this is one binary per release, per night, then our annual cost is ~$28, using crude arithmetic `(4* 100 * 365) / 1024) * 0.2`. So, it's not that expensive :)

File: build/teamcity/cockroach/nightlies/roachtest_compile_bits.sh
Comment (srosenberg) [reply build/teamcity/cockroach/nightlies/roachtest_compile_bits.sh:5]: Right. It'd be nice to export `root` or something similar. I find the other pattern (i.e., `dirname`) to be error-prone, especially when refactoring.

File: build/teamcity/cockroach/nightlies/roachtest_nightly_impl.sh
Comment (srosenberg) [reply build/teamcity/cockroach/nightlies/roachtest_nightly_impl.sh:62]: Nice catch! Forgot to uncomment after streamlining the process of testing in CI :)

File: build/teamcity/cockroach/nightlies/roachtest_compile_bits.sh
Comment (rickystewart) [reply build/teamcity/cockroach/nightlies/roachtest_compile_bits.sh:5]: I would add a comment like ("must run `build/teamcity-support.sh` before executing this script") at the top here just to make the dependency clear, then.

File: build/teamcity/util/roachtest_util.sh
Comment (srosenberg) [reply build/teamcity/util/roachtest_util.sh:71]: Updated the commit msg. I'll do a quick poll to see how many would prefer that mac and arm64 is also available.

File: build/teamcity/util/roachtest_util.sh
Comment (srosenberg) [reply build/teamcity/util/roachtest_util.sh:78]: > Or, even just one latest binary for master (per OS/arch combo)?

That was my original thought too. However, I opted for keeping nightly builds in case of regressions.

>  Realistically, people are not going to be downloading roachprod for older SHA's.

In most cases, that's true. When we do encounter a regression, it would be nice to tell folks, grab the binary with this SHA or wait until a fix is merged. The latter may take time, so it's nice to have a quick "workaround".

Comment (srosenberg) [reply build/teamcity/util/roachtest_util.sh:78]: Added a retention policy for binaries,

retention_policy

**Comment by srosenberg** [comment]
Everything is now green and ready to go. PTAL!

**Review by rickystewart — APPROVED** [review]

**Comment by srosenberg** [comment]
TFTR!

bors r=rickystewart

**Comment by renatolabs** [comment]
We can close #120750 now, yes? Or was there a reason this PR was only marked as `Informs`?

**Comment by rickystewart** [comment]
@renatolabs Check out the discussion [here](https://github.com/cockroachdb/cockroach/pull/124592#discussion_r1617780110).

**Pull request merged**
